sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -617,6 +618,46 @@ object NullPropagation extends Rule[LogicalPlan] {
 }
 
 /**
+ * Propagate foldable expressions:
+ * Replace all attributes with aliases of the original foldable expressions except Union queries.
+ * Aliases and ordinal expressions are the main target to be transformed after propagation. Other
+ * optimizations will take advantage of the propagated foldable expressions.
+ *
+ * {{{
+ *   SELECT 1.0 x, 'abc' y, Now() z ORDER BY x, y, 3
+ *   ==>  SELECT 1.0 x, 'abc' y, Now() z ORDER BY 1.0, 'abc', Now()
+ * }}}
+ */
+object FoldablePropagation extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = {
+    val isFoldableStatement = plan.find {
+      case _: Union => true
+      case _: Command => true
+      case _ => false
+    }.isEmpty
+
+    val foldableExprSet = ExpressionSet(plan.flatMap {
+      case Project(projectList, _) => projectList.collect {
+        case a: Alias if a.resolved && a.child.foldable => a
+      }
+      case _ => Nil
+    })
+
+    if (!isFoldableStatement || foldableExprSet.isEmpty) {
```

Comment:
Thank you, @cloud-fan !
I will try to add more fine-grained logic for them tomorrow.